Container locking - restrict container to currently assigned domains

mckenfra opened the pull request:
#847 Provides functionality for locking a container. Once locked, any requests for domains that are not in the assignments list for that container will open in a new tab.

Unit tests included.

<img width="305" alt="Container Unlocked" src="https://user-images.githubusercontent.com/439593/61718099-2cb17500-ad5a-11e9-9221-e0269a62b72d.png">

<img width="304" alt="Container Locked" src="https://user-images.githubusercontent.com/439593/61718110-3044fc00-ad5a-11e9-9af8-77df3035e8c7.png">

(This pull request is identical to the earlier one I submitted, but that one got cancelled because I moved the checkins to a separate branch on my fork.)

jonathanKingston commented:
@maxxcrawford also before we merge it's probably worth confirming with product or Tanvi we are good to merge as it's a fair amount of functional change.

mckenfra commented:
> Can you confirm that the icons (SVGs) in this PR are open-source and match the license used in this project?

The 2 new icons I added in my commit are taken (unchanged) from the MaterialUI project, github links here:

https://github.com/mui-org/material-ui/blob/9259aaaebbeded8b9b538a37c4be6094baede348/packages/material-ui-icons/src/Lock.js
https://github.com/mui-org/material-ui/blob/9259aaaebbeded8b9b538a37c4be6094baede348/packages/material-ui-icons/src/LockOpen.js

The licence used by the MaterialUI project is the MIT licence, github link here:

https://github.com/mui-org/material-ui/blob/9259aaaebbeded8b9b538a37c4be6094baede348/LICENSE

I believe the MIT licence is more permissive than Mozilla Public Licence 2.0 used by `mozilla/multi-account-containers`. It's certainly shorter! But the only issue I can see is that the MIT licence states:

> The above copyright notice and this permission notice shall be included in all
> copies or substantial portions of the Software.

Perhaps I need to actually insert the MIT licence as a comment at the top of both of the new SVG icon files. What do you think?

mckenfra commented:
@maxxcrawford @jonathanKingston I have updated the README to explain that the 2 newly-added icons are covered by the MIT licence, and have also added a copy of the MIT licence to the project directory. I believe the MIT licence is compatible with this project's licence.

maxxcrawford commented:
@mckenfra We already have a License file (https://github.com/mozilla/multi-account-containers/blob/master/LICENSE.txt), uses Mozilla's MPL 2.0 

@TanviHacks Can you review this PR too? Wanted to make sure this feature works for the roadmap.

groovecoder commented:
@mckenfra - can you try to resolve the conflicts here? We're considering merging this. Going to loop our UX folks into this too.

mckenfra commented:
> @mckenfra We'd like to update this PR and potentially get it into a future M-A-C release. (We'll have to consult with our legal & UX teams first on the LICENSE issues and if this UX works with their plans for Containers UX in Firefox.)
> 
> Will you have time to make the requested changes and updates? Or would you like @kendallcorner to pick this up from you?

Thanks a lot for carefully reviewing my changes! Yup, I'm happy to go through your requested updates. Sure, makes sense about consulting with your other teams too.

mckenfra commented on src/js/background/backgroundLogic.js:
```diff
@@ -229,7 +245,9 @@ const backgroundLogic = {
         numberOfHiddenTabs: containerState.hiddenTabs.length,
-        numberOfOpenTabs: openTabs.length
+        numberOfOpenTabs: openTabs.length,
+        // https://github.com/mozilla/multi-account-containers/issues/847
```
Sure, makes sense

mckenfra replied:
Done: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

mckenfra commented on src/js/popup.js:
```diff
@@ -980,6 +985,38 @@ Logic.registerPanel(P_CONTAINER_EDIT, {
+      lockElement.classList.add("container-info-tab-row", "clickable", "container-lockorunlock", lockOrUnlockClass);
+      tableElement.appendChild(lockElement);
+      
+      const that = this;
```
Well spotted! Yes I agree, there's no need for a `that` constant here. I'll change it to use `this`.

mckenfra replied:
Done: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

mckenfra commented on src/js/background/assignManager.js:
```diff
@@ -141,8 +141,21 @@ const assignManager = {
+    // This request's URL is not associated with any particular contextualIdentity.
+    // But the current tab's contextualIdentity is locked. So must open request in new tab.
+    // https://github.com/mozilla/multi-account-containers/issues/847
+    let isLockedOut;
```
Yes, I can rename all of the variable and function names from lock/unlock to something else. But my UI icons are very much lock/unlock symbols. Are you thinking of changing these too? Might it get a bit confusing if the icons are locks, but the code has references to "strict/relaxed"? But I do agree that getting the naming right in the code is important, so it's clear for future devs.

mckenfra replied:
Done: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

mckenfra commented on src/js/background/assignManager.js:
```diff
@@ -327,7 +347,20 @@ const assignManager = {
+          userContextId: userContextId,
+          isLocked: false
+        });
+      }
```
Done: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

mckenfra commented on src/js/background/backgroundLogic.js:
```diff
@@ -123,6 +123,22 @@ const backgroundLogic = {
+    }
+    return await identityState.storageArea.set(cookieStoreId, containerState);
+  },
+  
```
Shall we keep the locked/unlocked terminology? See improvements in: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

mckenfra commented on src/js/background/messageHandler.js:
```diff
@@ -16,6 +16,10 @@ const messageHandler = {
         response = backgroundLogic.createOrUpdateContainer(m.message);
         break;
+      case "lockOrUnlockContainer":
+        // https://github.com/mozilla/multi-account-containers/issues/847
```
Shall we keep the locked/unlocked terminology? See improvements in: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

mckenfra commented on src/js/popup.js:
```diff
@@ -4,6 +4,9 @@
 
 const CONTAINER_HIDE_SRC = "/img/container-hide.svg";
 const CONTAINER_UNHIDE_SRC = "/img/container-unhide.svg";
+// https://github.com/mozilla/multi-account-containers/issues/847
```
Done: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

mckenfra commented on src/js/popup.js:
```diff
@@ -252,6 +255,8 @@ const Logic = {
         identity.hasHiddenTabs = stateObject.hasHiddenTabs;
         identity.numberOfHiddenTabs = stateObject.numberOfHiddenTabs;
         identity.numberOfOpenTabs = stateObject.numberOfOpenTabs;
+        // https://github.com/mozilla/multi-account-containers/issues/847
```
Done: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

mckenfra commented on src/js/popup.js:
```diff
@@ -969,7 +974,7 @@ Logic.registerPanel(P_CONTAINER_EDIT, {
   },
 
-  showAssignedContainers(assignments) {
+  showAssignedContainers(assignments, isLocked) {
```
Shall we keep the locked/unlocked terminology? See improvements in: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

mckenfra commented:
> @mckenfra - can you try to resolve the conflicts here? We're considering merging this. Going to loop our UX folks into this too.

I'll resolve the conflicts as the last step, once you're happy with the other changes. Does that sound good?

mckenfra commented on src/js/popup.js:
```diff
@@ -252,6 +255,8 @@ const Logic = {
         identity.numberOfHiddenTabs = stateObject.numberOfHiddenTabs;
         identity.numberOfOpenTabs = stateObject.numberOfOpenTabs;
+        // https://github.com/mozilla/multi-account-containers/issues/847
+        identity.isLocked = stateObject.isLocked;
```
Shall we keep the locked/unlocked terminology? See improvements in: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

mckenfra commented:
> > @mckenfra - can you try to resolve the conflicts here? We're considering merging this. Going to loop our UX folks into this too.
> 
> I'll resolve the conflicts as the last step, once you're happy with the other changes. Does that sound good?

Done

kendallcorner commented:
Hi @mckenfra. This code has some merge conflicts with master right now, and I've even been working on some UX updates that (hopefully) should get merged soon that will create even more conflicts. *But* I *really* wanted to merge in this PR, so what I would like to do is copy the relevant code from this PR into my UX update PR and add you as a co-author. Would that be ok?

mckenfra commented:
> Hi @mckenfra. This code has some merge conflicts with master right now, and I've even been working on some UX updates that (hopefully) should get merged soon that will create even more conflicts. _But_ I _really_ wanted to merge in this PR, so what I would like to do is copy the relevant code from this PR into my UX update PR and add you as a co-author. Would that be ok?

@kendallcorner  Sure, go ahead!

kendallcorner commented:
Hi @mckenfra. This code has some merge conflicts with master right now, and I've even been working on some UX updates that (hopefully) should get merged soon that will create even more conflicts. *But* I *really* wanted to merge in this PR, so what I would like to do is copy the relevant code from this PR into my UX update PR and add you as a co-author. Would that be ok?

rubaboo commented:
So how's it going, guys? MAC will make much more sense when this functionality is added. Thank you.

JonRiv commented:
> So how's it going, guys? MAC will make much more sense when this functionality is added. Thank you.

+1 this is a most wanted feature ; without it behavior is a bit crippled if you actually want to protect cookies from your current container.

joweiser commented:
Only one last nit pick to fix over here: https://github.com/mozilla/multi-account-containers/pull/1654
Looks like this will be merged very soon - thanks @mckenfra and @kendallcorner for your work on this!
